datafusion/optimizer/src/optimizer.rs

```diff
@@ -422,6 +424,11 @@ fn assert_schema_is_the_same(
     prev_plan: &LogicalPlan,
     new_plan: &LogicalPlan,
 ) -> Result<()> {
+    if rule_name == "unnest_expressions" {
```

Comment:
There is the schema mismatch bug, I thought it is fine for `unnest`, let me take a look carefully.

Reply:
External error: query failed: DataFusion error: unnest_expressions
caused by
Internal error: Optimizer rule 'unnest_expressions' failed, due to generate a different schema, original schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(1),Int64(2),Int64(3)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(4),Int64(5)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: "sq0" }), field: Field { name: "col0", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: "sq1" }), field: Field { name: "col1", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
[SQL] select unnest(make_array(1,2,3)),
       unnest(make_array(4,5))
;

Reply:
Two differences in the schema.
1. qualifier with sq0, sq1..., since I wrap each unnest column with subquery
2. column name, I rename them to col0, col1..., just for simplicity.

@alamb Does it makes sense to skip these two checks and only check the other?
If not, is there an existing API to change the schema for LogicalPlan, so I can set None to qualifier and create a new field with the original name?

Reply:
let me check if those two are necessary or not.

Reply:
Internal error: Optimizer rule 'unnest_expressions' failed, due to generate a different schema, original schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(1),Int64(2),Int64(3)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "unnest(make_array(Int64(4),Int64(5)))", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "make_array(Int64(1),Int64(2),Int64(3))", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "make_array(Int64(4),Int64(5),NULL)", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }

I don't know how to fix these schema changes

name changes: unnest(make_array(Int64(1),Int64(2),Int64(3))) -> make_array(Int64(1),Int64(2),Int64(3))
data_type changes: List(int64) -> int64

Reply:
Need to review again

Reply:
https://github.com/apache/arrow-datafusion/blob/85ec31433615735a05d332f87cd0bdfc11aac663/datafusion/core/src/physical_plan/projection.rs#L423-L440

All nulls to align the row size at L432.